src/agent/state/oracle.rs

```diff
@@ -0,0 +1,618 @@
+    /// single request. Tune this setting if you're experiencing
+    /// timeouts on data fetching. In order to keep concurrent open
+    /// socket count at bay, the batches are looked up sequentially,
+    /// trading off overall time it takes to fetch all symbols.
+    pub max_lookup_batch_size: usize,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            commitment:               CommitmentLevel::Confirmed,
+            poll_interval_duration:   Duration::from_secs(5),
+            subscriber_enabled:       true,
+            updates_channel_capacity: 10000,
+            data_channel_capacity:    10000,
+            max_lookup_batch_size:    100,
+        }
+    }
+}
+
+pub struct OracleState {
+    data: RwLock<Data>,
+}
+
+impl OracleState {
+    pub fn new() -> Self {
+        Self {
+            data: Default::default(),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+pub trait Oracle {
+    async fn sync_global_store(&self, network: Network) -> Result<()>;
+    async fn poll_updates(
+        &self,
+        mapping_key: Pubkey,
+        rpc_client: &RpcClient,
+        max_lookup_batch_size: usize,
+        publisher_permissions_tx: Sender<HashMap<Pubkey, HashMap<Pubkey, PricePublishingMetadata>>>,
+    ) -> Result<()>;
+    async fn handle_price_account_update(
+        &self,
+        network: Network,
+        account_key: &Pubkey,
+        account: &Account,
+    ) -> Result<()>;
+}
+
+// Allow downcasting State into Keypairs for functions that depend on the `Keypairs` service.
+impl<'a> From<&'a State> for &'a OracleState {
+    fn from(state: &'a State) -> &'a OracleState {
+        &state.oracle
+    }
+}
+
+#[async_trait::async_trait]
+impl<T> Oracle for T
+where
+    for<'a> &'a T: Into<&'a OracleState>,
+    T: Prices,
+    T: Sync,
+{
+    async fn handle_price_account_update(
+        &self,
+        network: Network,
+        account_key: &Pubkey,
+        account: &Account,
+    ) -> Result<()> {
+        tracing::debug!("Handling account update.");
+
+        // We are only interested in price account updates, all other types of updates
+        // will be fetched using polling.
+        if !self
+            .into()
+            .data
+            .read()
+            .await
+            .price_accounts
+            .contains_key(account_key)
+        {
+            return Ok(());
+        }
+
+        let price_entry = PriceEntry::load_from_account(&account.data)
+            .with_context(|| format!("load price account {}", account_key))?;
+
+        tracing::debug!(
+            pubkey = account_key.to_string(),
+            price = price_entry.agg.price,
+            conf = price_entry.agg.conf,
+            status = ?price_entry.agg.status,
+            "Observed on-chain price account update.",
+        );
+
+        self.into()
+            .data
+            .write()
+            .await
+            .price_accounts
+            .insert(*account_key, price_entry.clone());
```

Comment:
i recall that there were some logics to make sure we only store the latest price, maybe you removed it because it's always more recent?

Reply:
my bad it's somewhere else. sorry.